docs/source/more_types.rst

```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+                   e: int, f: int, g: int, h: int) -> IPv6Adress:
+        pass
+
+    # The actual *implementation* of 'ip_address'.
+    # The implementation contains the actual runtime logic.
+    #
+    # It may or may not have type hints. If it does, mypy
+    # will check the body of the implementation against the
+    # type hints.
+    #
+    # Mypy will also check and make sure the signature is
+    # consistent with the provided variants.
+
+    def ip_address(*components: int) -> Union[IPv4Address, IPv6Address]:
+        if len(components) == 4:
+            # Return an IPv4Address object
+        elif len(components) == 8:
+            # Return an IPv6Address object
+        else:
+            # Raise an exception
+
+This allows mypy to understand calls to ``ip_address`` much more precisely.
+For example, mypy will understand that ``ip_address(127, 0, 0, 1)`` will
+always have a return type of ``IPv4Address`` and will report errors for
+calls like ``ip_address(1, 2)``.
+
+As another example, suppose we want to write a custom container class that
+implements the ``__getitem__`` method (``[]`` bracket indexing). If this
+method receives an integer we return a single item. If it receives a
+``slice``, we return a ``Sequence`` of items.
+
+We can precisely encode this relationship between the argument and the
+return type by using overloads like so:
+
+.. code-block:: python
+
+    from typing import Sequence, TypeVar, Union
+
+    T = TypeVar('T')
+
+    class MyList(Sequence[T]):
+        @overload
+        def __getitem__(self, index: int) -> T: pass
+
+        @overload
+        def __getitem__(self, index: slice) -> Sequence[T]: pass
+
+        def __getitem__(self, index: Union[int, slice]) -> Union[T, Sequence[T]]:
+            if isinstance(index, int):
+                # Return a T here
+            elif isinstance(index, slice):
+                # Return a sequence of Ts here
+            else:
+                raise TypeError(...)
+
+There are a few additional things to note about using overloads:
+
+Runtime behavior
+----------------
+
+An overloaded function must consist of two or more overload *variants*
+followed by an *implementation*. The variants and the implementations
+must be adjacent in the code: think of them as one indivisible unit.
+
+The variant bodies must all be empty; only the implementation is allowed
+to contain code. This is because at runtime, the variants are completely
+ignored: they're overridden by the final implementation function.
+
+This means that an overloaded function is still an ordinary Python
+function! There is no automatic dispatch handling: you must manually
+handle the different types in the implementation (usually by using
+``if`` statements and ``isinstance`` checks).
+
+If you are adding an overload within a stub file, the implementation
+function should be omitted: stubs do not contain runtime logic.
+
+Type checking calls to overloads
+--------------------------------
```

Comment:
I would place the section about checking definitions first. I think it is more logical, one first writes a function, then one calls it.

Reply:
> And the examples above already hint at how the definitions are checked.

This way the logic actually jumps IMO. We start from definitions in introduction, then we switch to checking calls, then then back to definitions. But this is not something critical, I can live with current ordering :-)

Reply:
Such jumping logic is often more readable than an exhaustive treatment per sub-topic -- this is probably different from how you've learned to write mathematical proofs. :-)